velox/exec/HashTable.cpp

```diff
@@ -1731,8 +1734,21 @@ template <bool ignoreNullKeys>
 void HashTable<ignoreNullKeys>::prepareJoinTable(
     std::vector<std::unique_ptr<BaseHashTable>> tables,
     int8_t spillInputStartPartitionBit,
-    folly::Executor* executor) {
+    folly::Executor* executor,
+    bool dropDuplicates) {
   buildExecutor_ = executor;
+  if (dropDuplicates) {
+    if (table_ != nullptr) {
+      // Set table_ to nullptr to trigger rehash.
+      rows_->pool()->freeContiguous(tableAllocation_);
+      table_ = nullptr;
```

Comment:
Is it necessary to clean up `capacity_ ` and `hashMode_` variables?

Reply:
No need to clean up other variables. Setting `table_ = nullptr` here is to allow entering the rehash process in the checkSize method, and it is not intended for data cleanup.

Reply:
```cpp
TEST_F(HashJoinTest, semiJoinDeduplicateResetCapacity) {
  // The initial size of HashTable is 2048. After detecting the
  // kAbandonBuildNoDupHashMinRows row, it is found that the duplication rate is
  // lower than kAbandonBuildNoDupHashMinPct, so no more data is inserted into
  // HashTable (the capacity_ variable does not change afterwards). However,
  // data will continue to be added to RowContainer, causing numDistinct_ to be
  // greater than 2048, and eventually an error is reported in the
  // HashTable<ignoreNullKeys>::checkSize function.
  const int vectorSize = 10, batches = 210;
  auto probeVectors = makeBatches(batches, [&](int32_t /*unused*/) {
    return makeRowVector({
        // Join Key is double -> VectorHasher::typeKindSupportsValueIds will
        // return false -> HashMode is kHash
        makeFlatVector<double>(
            vectorSize, [&](vector_size_t /*row*/) { return rand(); }),
        makeFlatVector<int64_t>(
            vectorSize, [&](vector_size_t /*row*/) { return rand(); }),
    });
  });

  auto buildVectors = makeBatches(batches, [&](int32_t batch) {
    return makeRowVector({
        makeFlatVector<double>(
            vectorSize, [&](vector_size_t /*row*/) { return rand(); }),
        makeFlatVector<int64_t>(
            vectorSize, [&](vector_size_t /*row*/) { return rand(); }),
    });
  });

  createDuckDbTable("t", probeVectors);
  createDuckDbTable("u", buildVectors);

  auto planNodeIdGenerator = std::make_shared<core::PlanNodeIdGenerator>();
  auto plan = PlanBuilder(planNodeIdGenerator)
                  .values(probeVectors)
                  .project({"c0 AS t0", "c1 AS t1"})
                  .hashJoin(
                      {"t0"},
                      {"u0"},
                      PlanBuilder(planNodeIdGenerator)
                          .values(buildVectors)
                          .project({"c0 AS u0", "c1 AS u1"})
                          .planNode(),
                      "",
                      {"t0", "t1", "match"},
                      core::JoinType::kLeftSemiProject)
                  .planNode();

  HashJoinBuilder(*pool_, duckDbQueryRunner_, driverExecutor_.get())
      .config(core::QueryConfig::kAbandonBuildNoDupHashMinRows, "10")
      .config(core::QueryConfig::kAbandonBuildNoDupHashMinPct, "50")
      .numDrivers(1)
      .checkSpillStats(false)
      .planNode(plan)
      .referenceQuery(
          "SELECT t.c0, t.c1, EXISTS (SELECT * FROM u WHERE t.c0 = u.c0) FROM t")
      .run();
}
```
This case will not work without resetting the capacity variable.

Reply:
@XinShuoWang Thank you for your review. I understand that when "abandon" is triggered, the `capacity_` is not accurate and needs to be reset. I will make the necessary modifications and incorporate your test cases.